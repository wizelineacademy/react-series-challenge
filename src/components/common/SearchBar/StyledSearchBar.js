import React, { Component } from 'react'
import { connect } from 'react-redux';
import styled from 'styled-components'
import gifs from '../../../actions/gifs'
import searchicon from './searchicon.png'


const Input = styled.input`
    width: 300px;
    margin:0 auto;
    box-sizing: border-box;
    border: 2px solid #ccc;
    font-weight:bold;
    border-radius: 4px;
    outline:none;
    border-radius:1em;
    font-size: 16px;
    background-color: white;
    background-image: url(${searchicon});
    background-position: 10px 10px; 
    background-repeat: no-repeat;
    padding: 12px 20px 12px 40px;
    -webkit-transition: width 0.4s ease-in-out;
    transition: width 0.4s ease-in-out;
    &:focus{
        width: 100%;
        background-color:rgba(63,191,127,0.8);
        color:#fff;
        
    }
`

class SearchBar extends Component{
    onChangeSearch(e){
        this.props.setQuery(e.target.value)
    }
    render(){
        return(
            <form onSubmit={
                (e)=>{
                    e.preventDefault();
                   
                   const {query, searchFunction} = this.props
                   
                   searchFunction(query)
                }
            }>
                <Input onChange={this.onChangeSearch.bind(this)} value={this.props.query} type="text" placeholder="Buscar (Enter)"/>
            </form>
       )
    }
}
const querySelector = state =>{
    return state.giphyReducer.query
}
const mapStateTopProps = state =>({
    query:querySelector(state)
})
const { creators } = gifs
const { setQuery } = creators
const mapDispatchToProps = dispatch =>({
   
    setQuery(query){
        dispatch(setQuery(query))
    },
})

export default connect(mapStateTopProps,mapDispatchToProps)(SearchBar)