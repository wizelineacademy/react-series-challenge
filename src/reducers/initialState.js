export default {
  searchQuery: '',
  filterFavorites: '',
  searchResults: [],
  trendingGifs: [],
  favorites: []
};