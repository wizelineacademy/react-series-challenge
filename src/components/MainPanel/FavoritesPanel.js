import React, { Component } from 'react';
import GifListView from "./GifListView";
import ErrorBoundary from "../ErrorBoundary";

import actions from "../../actions";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";

import StyledSearchBar from '../styled/SearchBar.styled';

const { filterFavorites } = actions.creators;

const FavoritesPanel = ({ data, filterFavorites, filter }) => {
  const filteredData = data.filter((element) => (
    element.title.toLowerCase().includes(filter.toLowerCase())
  ))
  return (
    <div>
      <StyledSearchBar 
        type="text" 
        placeholder="Filter" 
        onChange={(e) => filterFavorites(e.target.value)}
      />
      <ErrorBoundary message="Unable to load.">
        <GifListView data={filteredData} />
      </ErrorBoundary>
    </div>
  )
}

const mapStateToProps = state => {
  return {
    data: state.favorites,
    filter: state.filterFavorites
  }
}

const mapDispatchToProps = dispatch => (
  bindActionCreators({ filterFavorites }, dispatch)
)

export default connect(mapStateToProps, mapDispatchToProps)(FavoritesPanel);