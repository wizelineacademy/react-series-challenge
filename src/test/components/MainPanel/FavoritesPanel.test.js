import React from 'react';
import { shallow, render, mount } from "enzyme";
import { Provider } from 'react-redux';

import { mockGif } from "../../mocks";
import initialState from "../../../reducers/initialState";

import FavoritesPanel from "../../../components/MainPanel/FavoritesPanel";
import StyledSearchBar from '../../../components/styled/SearchBar.styled';

describe('FavoritesPanel', () => {
  
  it('FavoritesPanel Component should match snapshot (WITHOUT filter)', () => {
    const mockStore = {
      dispatch: jest.fn(),
      getState: jest.fn(() => ({
        ...initialState,
        favorites: [mockGif]
      })),
      subscribe: jest.fn()
    }
    expect(shallow(<FavoritesPanel store={mockStore} />)).toMatchSnapshot();
  });

  it('FavoritesPanel Component should match snapshot (WITH filter)', () => {
    const mockStore = {
      dispatch: jest.fn(),
      getState: jest.fn(() => ({
        ...initialState,
        favorites: [mockGif],
        filterFavorites: 't'
      })),
      subscribe: jest.fn()
    }
    expect(render(
      <Provider store={mockStore} >
        <FavoritesPanel/>
      </Provider>
    )).toMatchSnapshot();
  });

  it('Input change should dispatch action', () => {
    const mockStore = {
      dispatch: jest.fn(),
      getState: jest.fn(() => ({
        ...initialState,
        favorites: [mockGif]
      })),
      subscribe: jest.fn()
    }
    const comp = mount(
      <Provider store={mockStore} >
        <FavoritesPanel/>
      </Provider>
    );
    comp.find(StyledSearchBar).simulate('change');
    expect(mockStore.dispatch).toHaveBeenCalled();
  });

})
