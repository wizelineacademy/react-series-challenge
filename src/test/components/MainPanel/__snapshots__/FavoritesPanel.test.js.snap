// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritesPanel FavoritesPanel Component should match snapshot (WITH filter) 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "class": "sc-ifAKCX gobjtK",
          "placeholder": "Filter",
          "type": "text",
        },
        "children": Array [],
        "name": "input",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "sc-bxivhb ckxRMc",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "sc-htpNat eTHDIu",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "sc-bwzfXH fzZGbX",
                    "src": "url",
                  },
                  "children": Array [],
                  "name": "img",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": Object {
                    "attribs": Object {
                      "class": "sc-bdVaJa owqAu",
                      "src": "addToFavorites.png",
                    },
                    "children": Array [],
                    "name": "img",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": [Circular],
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                      "src": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                      "src": undefined,
                    },
                  },
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                    "src": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                    "src": undefined,
                  },
                },
                Object {
                  "attribs": Object {
                    "class": "sc-bdVaJa owqAu",
                    "src": "addToFavorites.png",
                  },
                  "children": Array [],
                  "name": "img",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": Object {
                    "attribs": Object {
                      "class": "sc-bwzfXH fzZGbX",
                      "src": "url",
                    },
                    "children": Array [],
                    "name": "img",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": [Circular],
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                      "src": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                      "src": undefined,
                    },
                  },
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                    "src": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                    "src": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "placeholder": undefined,
          "type": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "placeholder": undefined,
          "type": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "sc-bxivhb ckxRMc",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "sc-htpNat eTHDIu",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "sc-bwzfXH fzZGbX",
                  "src": "url",
                },
                "children": Array [],
                "name": "img",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": Object {
                  "attribs": Object {
                    "class": "sc-bdVaJa owqAu",
                    "src": "addToFavorites.png",
                  },
                  "children": Array [],
                  "name": "img",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": [Circular],
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                    "src": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                    "src": undefined,
                  },
                },
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                  "src": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                  "src": undefined,
                },
              },
              Object {
                "attribs": Object {
                  "class": "sc-bdVaJa owqAu",
                  "src": "addToFavorites.png",
                },
                "children": Array [],
                "name": "img",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": Object {
                  "attribs": Object {
                    "class": "sc-bwzfXH fzZGbX",
                    "src": "url",
                  },
                  "children": Array [],
                  "name": "img",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": [Circular],
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                    "src": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                    "src": undefined,
                  },
                },
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                  "src": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                  "src": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "sc-ifAKCX gobjtK",
            "placeholder": "Filter",
            "type": "text",
          },
          "children": Array [],
          "name": "input",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "placeholder": undefined,
            "type": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "placeholder": undefined,
            "type": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`FavoritesPanel FavoritesPanel Component should match snapshot (WITHOUT filter) 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(FavoritesPanel)
    store={
        Object {
            "dispatch": [Function],
            "getState": [Function],
            "subscribe": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "data": Array [
        Object {
          "favorite": false,
          "id": "id",
          "original": "original",
          "title": "title",
          "url": "url",
        },
      ],
      "filter": "",
      "filterFavorites": [Function],
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": undefined,
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "data": Array [
          Object {
            "favorite": false,
            "id": "id",
            "original": "original",
            "title": "title",
            "url": "url",
          },
        ],
        "filter": "",
        "filterFavorites": [Function],
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "dispatch": [Function],
            "getState": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": undefined,
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
